File: plugins/plugin-native-mobile-signals/android/src/main/java/ai/eliza/plugins/mobilesignals/MonitoringGeneration.kt
/**
 * Monotonic ownership token for asynchronous native monitoring work.
 *
 * A completion may publish only while its captured generation is current, so
 * stopping and restarting cannot make an old coroutine look active again.
 */
package ai.eliza.plugins.mobilesignals

import java.util.concurrent.atomic.AtomicLong

internal class MonitoringGeneration {
    private val current = AtomicLong(0)

    fun begin(): Long = current.incrementAndGet()

    fun invalidate(): Long = current.incrementAndGet()

    fun isCurrent(generation: Long): Boolean = current.get() == generation
}

File: plugins/plugin-native-mobile-signals/ios/Sources/MobileSignalsHealthContract/MonitoringGeneration.swift
/**
 * Monotonic ownership token for asynchronous native monitoring work.
 *
 * Completions publish only while their captured generation is current, so a
 * stopped HealthKit read cannot become active again after an immediate restart.
 */
import Foundation

public final class MonitoringGeneration {
    private let lock = NSLock()
    private var current: UInt64 = 0

    public init() {}

    @discardableResult
    public func begin() -> UInt64 {
        lock.lock()
        defer { lock.unlock() }
        current &+= 1
        return current
    }

    @discardableResult
    public func invalidate() -> UInt64 {
        lock.lock()
        defer { lock.unlock() }
        current &+= 1
        return current
    }

    public func isCurrent(_ generation: UInt64) -> Bool {
        lock.lock()
        defer { lock.unlock() }
        return current == generation
    }
}
